docs/ABI/RegisterAllocation.md

```diff
@@ -0,0 +1,13 @@
+## 64 bit Architecture Register Allocations
+
+From Swift 5, the calling convention register allocation for 64 architectures is largely based on [existing](https://developer.apple.com/library/content/documentation/Xcode/Conceptual/iPhoneOSABIReference/Articles/ARM64FunctionCallingConventions.html) [standards](https://developer.apple.com/library/content/documentation/DeveloperTools/Conceptual/LowLevelABI/140-x86-64_Function_Calling_Conventions/x86_64.html) with the addition of the context and error return registers discussed in the ABI manifesto:
+
+| Register Purpose | ARM64 | x86_64 |
+| ------------- |:-------------:| ----- |
+| Context register (self) | x20 | r13 |
+| Error return register | x21 | r12 |
+| Struct return pointer | x8 | rax |
+| Float call arguments | d0 - d7 | xmm0 - xmm7 |
+| Integer call arguments | x0 - x7 | rdi, rsi, rdx, rcx, r8, r9 |
+| Float return | d0, d1 | xmm0, xmm1 |
+| Integer return | x1, x2 | rax, rbx |
```

Comment:
IIRC we use (rax, rdx, rcx, r8) for integer returns on x86, and x0...x3 on arm64.

Reply:
Interesting... I've mostly had to reverse engineer this. It really needs to be documented! I was using it in [this project](https://github.com/johnno1962/SwiftTrace) which you may appreciate but will likely have you shaking your head.

Reply:
You might be interested in the `@_dynamicReplacement` feature that's being developed in the master branch. What you're doing looks like it could be implemented in a better-supported way using it.

Reply:
Very interesting... What's the PR?

Reply:
Excellent, you've solved the last problem I was having on x86. You can now add an aspect to cancel a throw and provide a value 🤦

Reply:
One question - does the ABI use d0..d3/xmm0..xmm3 for floats?

Reply:
Seems so

Reply:
No, floats use v0...v7, and xmm0...xmm7, just like doubles. They just only use the low 32b of each register.

Reply:
> What's the PR?

It is committed to the master branch of the compiler.

Reply:
Ha, Swizzling lives! Are there really only 6 integer registers used for x86 arguments. I was going on the System V document.

Reply:
AFAICT yes, we use the standard argument registers, aside from the error and self registers.